### PR TITLE
API Update template lookup to late resolution for performance reasons

### PR DESCRIPTION
**Pull request by tractorcow**
This proposed update to templating behaviour addresses a few issues.

Broken out of https://github.com/silverstripe/silverstripe-framework/pull/5921

Merge with https://github.com/silverstripe/silverstripe-cms/pull/1598

*Performance*

As SSViewer::get_templates_by_class() invokes SSViewer::has_template on EVERY item in the chain (even if we have higher precedence templates) it will end up doing redundant work. This is because SSViewer::setTemplate will choose only the first template in any given list that exists within the set of themes.

To address this, instead of doing any pre-filtering, we simply generate the list of all candidate templates, regardless of whether they exist, and rely on setTemplate() to do the final candidate selection.

SSViewer::chooseTemplate() has been refactored out in order to support cases where you need to know if a valid template is available, or to force template choosing.

*Consistency*

Previously, `forms` wasn't a standard type, as `Includes` or `Content` types were. This means that SSviewer::get_templates_by_class() couldn't be used at all for formfields due to it only supporting a template suffix, and not a `forms/` prefix.

When form fields are templated, this would mean that we could have duplicate directories for each namespace (one under a top level `forms/` folder and one at root). This change in behaviour flips the `forms` back into a template type, so that we can use the same lookup inheritance chain to search for templates.

A lot of gridfield components with hard (and semi-hard) coded templates have been updated.

ClassInfo has also been updated to clarify where you can use an instance of an object in place of a class name.

GridFieldItemEditView.ss has been removed as in all cases, this template is overridden and is redundant.

### CODE REVIEW

**Comment by tractorcow** [comment]
It looks like this breaks the CMS in a weird way. the save button is gone in profile edit, and you can't scroll.

**Comment by tractorcow** [comment]
Regressions have been resolved; there was a missing 'forms' type required for custom formfield templates, and I wasn't respecting full templates being passed into ThemeResourceLoader::findTemplate() properly (new tests added).

**Comment by tractorcow** [comment]
I've updated this PR to remove the `forms` type from the template lookup.

Merge with https://github.com/silverstripe/silverstripe-cms/pull/1598.

**Comment by tractorcow** [comment]
Green and ready for merge. :)

**Comment by chillu** [comment]
Resolved merge conflict in changelog and force pushed

File: admin/code/SecurityAdmin.php [hunk 1]
```diff
@@ -165,7 +165,7 @@ public function getEditForm($id = null, $fields = null) {
 		// Add import capabilities. Limit to admin since the import logic can affect assigned permissions
 		if(Permission::check('ADMIN')) {
 			$fields->addFieldsToTab('Root.Users', array(
-				new HeaderField(_t('SecurityAdmin.IMPORTUSERS', 'Import users'), 3),
+				new HeaderField('ImportUsersHeader', _t('SecurityAdmin.IMPORTUSERS', 'Import users'), 3),
```
Comment (chillu) [inline admin/code/SecurityAdmin.php]: Technically not part of the PR topic (fix templates), but all good :)

File: api/RSSFeed.php [hunk 2]
```diff
@@ -240,6 +240,22 @@ public function setTemplate($template) {
 	public function getTemplate() {
 		return $this->template;
 	}
+
+	/**
+	 * Returs the ordered list of preferred templates for rendering this form
```
Comment (chillu) [inline api/RSSFeed.php]: The PHPDoc is misleading, it refers to `Form` in the `RSSFeed` class (and misspelled "Returns")

File: forms/gridfield/GridFieldDetailForm.php [hunk 3]
```diff
@@ -752,6 +752,21 @@ public function getTemplate() {
 	}
 
 	/**
+	 * Get list of templates to use
+	 *
+	 * @return array
+	 */
+	public function getTemplates()
+	{
+		$templates = SSViewer::get_templates_by_class($this, '', __CLASS__);
```
Comment (chillu) [inline forms/gridfield/GridFieldDetailForm.php]: Right, `__CLASS__` refers to the base class where the method context was defined (http://php.net/manual/en/language.constants.predefined.php). Was wondering if this breaks on subclasses which wouldn't look for templates matching parent classes, but we're good.

File: view/SSViewer.php [hunk 4]
```diff
@@ -861,49 +861,47 @@ public static function set_theme($theme) {
 	/**
 	 * Traverses the given the given class context looking for templates with the relevant name.
```
Comment (chillu) [inline view/SSViewer.php]: That description is a bit ambiguous now, we should clarify that it returns a list of potentially existing templates - it doesn't actually *look* for templates any more.

File: docs/en/04_Changelogs/4.0.0.md [hunk 5]
```diff
@@ -63,6 +63,7 @@
  * Search filter classes (e.g. `ExactMatchFilter`) are now registered with `Injector`
    via a new `DataListFilter.` prefix convention.
    see [search filter documentation](/developer_guides/model/searchfilters) for more information.
+ * FormField templates no longer look in the 'forms' folder for templates.
```
Comment (chillu) [inline docs/en/04_Changelogs/4.0.0.md]: Should mention the changed semantics of `SSViewer:: get_templates_by_class()` (returning all templates regardless of existence)

**Comment by chillu** [comment]
`Email@parseVariables`is still using `$template = new SSViewer('email/'.$this->ss_template)` - does that need to be fixed?

File: forms/gridfield/GridFieldDetailForm.php (hunk 3 above)
Comment (tractorcow) [inline forms/gridfield/GridFieldDetailForm.php]: Yeah, I've started using this extensively to save having to use literal class strings (which are undesirable due to their difficulty to detect and upgrade).

**Comment by tractorcow** [comment]
All above addressed. How does it look @chillu? Do the documentation updates mesh with your view? :D

File: docs/en/04_Changelogs/4.0.0.md (hunk 5 above)
Comment (tractorcow) [inline docs/en/04_Changelogs/4.0.0.md]: I've now added this a few lines down.

File: docs/en/04_Changelogs/4.0.0.md [hunk 6]
```diff
@@ -415,8 +419,14 @@ require manual intervention, please see the below upgrading notes.
 Templates are now much more strict about their locations. You can no longer put a template in an arbitrary
 folder and have it be found. Case is now also checked on case-sensitive filesystems.
 
-Either include the folder in the template name (`renderWith('Field.ss')` => `renderWith('forms/Field.ss')`), move the 
-template into the correct directory, or both.
+Either include the folder in the template name (`renderWith('MyEmail.ss')` => `renderWith('emails/MyEmail.ss')`),
+move the template into the correct directory, or both.
+
+When using `<% include %>` template tag you should continue to leave out the `Includes` folder.
```
Comment (chillu) [inline docs/en/04_Changelogs/4.0.0.md]: This advice is only accurate until we implement the suggestions in https://github.com/silverstripe/silverstripe-framework/issues/5952#issuecomment-244881303 (where `<% include Includes/Sidebar %>` will become the new and consistent default)

File: forms/gridfield/GridFieldPrintButton.php [hunk 7]
```diff
@@ -121,8 +121,10 @@ public function handlePrint($gridField, $request = null) {
 		Requirements::css(FRAMEWORK_DIR . '/client/dist/styles/GridField_print.css');
 
 		if($data = $this->generatePrintData($gridField)){
-			return $data->renderWith("GridField_print");
+			return $data->renderWith(get_class($gridField)."_print");
```
Comment (chillu) [inline forms/gridfield/GridFieldPrintButton.php]: Wouldn't this break on subclasses? `renderWith()` itself does not have a concept of a class hierarchy traversal to find "parent class templates", and `get_class()` always retrieves the concrete class of the instance.

How about this? I don't see a clean way to identify the "base class" of `GridField` without hardcoding here.
```
$templates = SSViewer::get_templates_by_class($gridField, '_print', 'GridField);
return $data->renderWith($templates);
```

Happy to merge either way, given that `GridField` is usually not extended directly it's a bit of an edge case.
